### ISSUE TITLE
Add support for sequence annotations

### ISSUE THREAD

**Issue by diekhans** [ga4gh/server#362]
We would like to request prioritizing support for the sequence annotation API.  The variant API plus the sequence annotation API will enable us to demonstrate the usefulness of the GA4GH APIs in the BRCA challenge.  We would like to be able to report on this at the Paris meeting.

The sequence annotations methods will probably need revamped.

**Comment by pgrosu** [comment]
It might not be a bad idea to start discussing ways of having the server team work on a standardized, scalable development environment on something like AWS (Amazon), Google or another cluster.  Maybe there is a way this can be provided via GA4GH, similarly as [initiated here](https://github.com/ga4gh/server/issues/335).  I think it will provide the opportunity not only to ensure that everyone is using the same setup environment - especially for deployment activities - but more importantly the exposure to experiment with services such as Elastic MapReduce, Amazon Redshift, Amazon S3, Google Cloud Storage, Google BigQuery, Google Cloud Dataflow, Google Compute Engine, as well as adding and customizing open-source software tools.

I feel this will be necessary for large-scale coordinated initiatives such as the BRCA Challenge, and as well as tackling and integrating the different annotation APIs.

Paul

**Comment by skeenan** [comment]
@pgrosu, @diekhans has requested that support for the sequence-annotation API is prioritised for the reference implementation.  I think your comments on deployment environments should be moved out to separate issue.

**Comment by pgrosu** [comment]
@skeenan, I would be glad to.  I was wondering what might be the best way to join the DWG mailing list and conference calls.  It might help me with being better in sync with the discussions regarding the APIs, besides just through the Issues/PRs on GitHub.

Thank you,
Paul

**Comment by benedictpaten** [comment]
@pgrosu - the team is developing S3 support and we at UCSC plan to put up a public test server on S3. I believe there are a couple of outstanding technical issues that are in the process of being solved. I agree with @skeenan - this is off topic for this thread.

**Comment by pgrosu** [comment]
@benedictpaten, so may I open up another issue that expands on this?  S3 is optimized for storage, but the server team I think would find great benefits of trying out the other services with other parts of this project - especially for indexing and searching.

**Comment by jeromekelleher** [comment]
@diekhans: enthusiastically agree!

**Comment by diekhans** [comment]
While the reference server should run in AWS, it will not be
optimized for AWS.  That would be counter to the reference
server goals of simplicity.  There is a need for servers highly
optimized for the cloud.

**Comment by dcolligan** [comment]
I've created an annotations branch in both the schemas and the server repositories.  We can work off of those.  (Just don't force push to them, or else you'll confuse everyone!)

https://github.com/ga4gh/schemas/tree/annotations

https://github.com/ga4gh/server/tree/annotations

**Comment by dcolligan** [comment]
Okay, we'll we've hit our first schema bug!  There's some problem with ReadAlignment.duplicateFragment: https://gist.github.com/dcolligan/45af0974aeaeb1396b6a

**Comment by dcolligan** [comment]
@macieksmuga do you have a branch open where a lot of these schema bugs are fixed?

**Comment by diekhans** [comment]
We need to setup CI for schemas!

**Comment by pgrosu** [comment]
Not a bad idea, but I think in this instance `schemas` is fine.  So by looking under the [Avro specs](https://avro.apache.org/docs/1.7.7/spec.html#Unions) for unions, you'll notice the following:

*(Note that when a default value is specified for a record field whose type is a union, the type of the default value must match the first element of the union. Thus, for unions containing "null", the "null" is usually listed first, since the default value of such unions is typically null.)*

So you have two options for the `server` implementation:

1) Make this check order-independent, by first pre-processing the schema on-the-fly and moving `null` at the beginning of all `union` instances, so this step will always work:

```Python
   if (isinstance(t0, avro.schema.PrimitiveSchema) and
                            t0.type == "null"):
```
or,

2) Implementing the Avro interpretation for unions where the default value matches the first element-type of the union.

Hope it helps,
Paul

**Comment by diekhans** [comment]
Plan is to write a program to convert GFF3 to the GA4GH sequence annotation JSON and a second program to load this into a sqlite3 database. 

Q: how is the best way to structure this in the tree, as each of these programs will have command line tests with data?

**Comment by pgrosu** [comment]
Since annotation will become very broad, it probably should have its own directory under:

https://github.com/ga4gh/server/tree/develop/ga4gh

With time we probably want several categorized directories encompassing the large general functionalities being provided (i.e. Data Processing, Annotation, Analysis, Import_Export, etc.) - it will make it more naturally to navigate hierarchically and interface among them (both visually and via code).

Paul

**Comment by macieksmuga** [comment]
I'm assuming that if null is the second element of a union, this is explicitly done to force the default value of the union to be something other than null. Pre-processing the schemas to change this order would thus be going against the spec. Unless there's a mistake in the schemas, and should be corrected there - per @dcolligan's hunch (I don't know enough about SAM flags to tell if the `union { boolean, null }` is intentional or not).

**Comment by macieksmuga** [comment]
@dcolligan I'm formulating several pull requests against the schemas repo to correct various minor issues that I've found. The specific one you encountered here, however, isn't among them (I was assuming this union order was intentional on the part of the schema authors, and that we should change the null-as-first assumption on our end). Here's a fairly well documented example of a non-null first element of a union from the `SearchReadsRequest` record in `readmethods.avdl`:
```
  /**
  The start position (0-based) of this query.
  If a reference is specified, this defaults to 0.
  Genomic positions are non-negative integers less than reference length.
  Requests spanning the join of circular genomes are represented as
  two requests one on each side of the join (position 0).
  */
  union { long, null } start = 0;
```
On the other hand, thinking about it some more I'm now kind of wondering why null *isn't* the default here (since `referenceId` defaults to null, in which case `start` should be null as well). So... yeah, maybe I should add it to my growing list of schemas PR's.

**Comment by diekhans** [comment]
Yes, the package structure needs to evolve.

However, right now I am concerned about how I organize the
command line program and tests that are completely independent
of the server.

**Comment by pgrosu** [comment]
@macieksmuga, don't change the `schemas` yet, it works fine according to the Avro definitions.  The `server` processing of Avro has to change.  As this definition from Avro for unions suggests, `null` does not need to be first:

*(Note that when a default value is specified for a record field whose type is a union, the type of the default value must match the first element of the union. Thus, for unions containing "null", the "null" is usually listed first, since the default value of such unions is typically null.)*

You can find this definition here:

https://avro.apache.org/docs/1.7.7/spec.html#Unions

Paul

**Comment by fnothaft** [comment]
> @macieksmuga, don't change the schemas yet, it works fine according to the Avro definitions. The server processing of Avro has to change. As this definition from Avro for unions suggests, null does not need to be first:

No, this is a known issue in the schemas that should be fixed in the schemas repository. Null doesn't need to be first, _but_ the default value must match the type listed first in the union.

**Comment by pgrosu** [comment]
@fnothaft, so if the default is `false` and `boolean` is the first element, why do we need change the schemas if this is requirement is followed?  The `server` team is not seeing `null` when processing it with the file below, and wants to change all the unions where `null` is not the first element to be the first element, because it fails here:

https://github.com/ga4gh/server/blob/develop/scripts/process_schemas.py#L99-104

```Python
                    if (isinstance(t0, avro.schema.PrimitiveSchema) and
                            t0.type == "null"):
                        if isinstance(t1, avro.schema.RecordSchema):
                            ret.append((field.name, t1.name))
                    else:
                        raise Exception("Schema union assumptions violated")
```

I feel the `server` code would need to be updated to be Avro-compliant, instead of changing the schema or data because it does not fit the code.

~p

**Comment by fnothaft** [comment]
@pgrosu there are multiple fields in the schemas repository that do not follow the correct conventions and that need to be fixed.

**Comment by pgrosu** [comment]
@diekhans If this will be initially an independent endeavor, just create a `Projects` folder under [https://github.com/ga4gh/server](https://github.com/ga4gh/server).  Then you can have it look something like this:

```
Projects /
  Annotation/
    GFF3_to_GA4GH-JSON
    GA4GH-JSON_to_sqlite3
```

Then within each one (starting under the Annotation-grouped projects level) you can go with something like this:

```
  Annotation /
     GFF3_to_GA4GH-JSON /
	 
	 
		Main /
			 build (or compiled) /
			 config /
			 data /
			 docs /
			 libs /
			 src /
			 tests /
			 
			 
		Development /
		     FeatureBranch1 /
			     build /
			     config /
				 data /
				 docs /
			     lib /
				 src /
				 tests /
					  
			 FeatureBranch2 /
			     build /
			     config /
				 data /
				 docs /
			     lib /
				 src /
				 tests /
			

		Releases /
		     Release1.1 /
				 build /
				 config /
				 data /
				 docs /
				 lib /
				 src /
					  
			 Release1.2 /
				 build /
			     src /
```

It doesn't have to be too structured as above, but it helps.  Maybe starting from a Maven-type directory structure might be the easiest to begin.  Over time I'm sure it'll evolve.  Once it's ready for staging, then options for integration with the rest of `server` can also explored.  Thus deployment would be most natural.

Hope it helps,
Paul

**Comment by pgrosu** [comment]
@fnothaft Agreed and I'm all for Continuous Integration (CI), but for places where the schema is designed correctly, it would be nice to make the code adapt to the data, rather the other way around.  This way consistency in following the standard is preserved across all project components.

**Comment by dcolligan** [comment]
In the case of `ReadAlignment.duplicateFragment` I'm unsure why this is a union type at all; a boolean type would seem more appropriate.  It's supposed to map to a SAM flag.  If `duplicateFragment` is false the flag is unset and if it is true the flag is set... but what is null supposed to mean?  I would have the same question for all other ternary flag attributes.

The `process_schemas.py` code in the same method where the exception is thrown does have this comment:
```
        # TODO need to clarify how we operate on Unions here. The current
        # code will break when we move to schema version 0.6 as we are
        # no longer assured that the first element of the union is null.
        # This would be a good opportunity to tidy this up.
```
How should we correctly process union types here?  I'm unsure what the intention was supposed to be.

**Comment by fnothaft** [comment]
@pgrosu I don't think anyone is saying to make the server code "override" the schemas. If I read the thread correctly, people are saying two things:

1. There are fields in the schemas that do not correctly follow avro rules for default typing, which I can confirm is a bug in the schema repository. To avoid these issues in the future, we should have better CI testing.
2. Some of the fields define non-null default values that are semantically incongruent with other default values in the same schema, or the fields should not be nullable. E.g., my reading of @macieksmuga's comment RE: `union { long, null } start = 0`, is that a default value of `0` for `start` doesn't make sense since the default value for `referenceId` is `null`, and that `start` is not meaningful unless `referenceId` is defined.

**Comment by macieksmuga** [comment]
Thank you @fnothaft for the clarification! I'm still predisposed to believe that things that puzzle me about the schemas have more to do with my lack of understanding, so it's nice to know that sometimes they're just simple mistakes.

**Comment by dcolligan** [comment]
@jeromekelleher wrote that comment in `process_schemas.py` so perhaps he could shed some light on what the correct thing to do is wrt union types?

**Comment by macieksmuga** [comment]
I second @dcolligan's question about unions with booleans, such as in `ReadAlignment.duplicateFragment`. Perhaps somebody well versed in SAM flag semantics could chime in on the design intent?

**Comment by pgrosu** [comment]
@fnothaft Depends on how it is interpreted.  So there are two scenarios in @macieksmuga example:

1) If `referenceId` is empty or `null`, then all the reads are returned from all references.  If one returns all reads from all references, then by default the list of references defines at least one reference.  If one defines at least one reference, then it satisfies *"If a reference is specified, this defaults to 0."*, thus setting `start` to 0.

2) If one provides a `referenceId`, then it satisfies "If a reference is specified, this defaults to 0.", thus setting `start` to 0.

So I'm fine either way with @macieksmuga changing it, but we need to implement a rule-based checker that is based on the interpretation of the comments in the `schemas` - as well as their definitions for each implemented record/type - which will run and validate all the scenarios we construct.  This would need to be updated with each PR in `schemas`.  This can involve multiple combinations of records and types as designed in the `schemas`.  

Usually when I construct programs, I follow a consistent design recipe where I first start with the test cases of how I expect the outputs to be based on the inputs, while using only the function names.  This takes up the majority of the time.  Then I add the function contracts to the functions (i.e. the defined arguments/parameters) in terms of inputs and outputs based on the tests.  I do that first to the tests and then the defined functions.  Then these define the commented definition for each function and its types - the interpretation was previously defined via the test cases.  It's only the last 10% that I actually code, which becomes a very mechanical process, since the plan is already written for me by me, having gone through the process of analyzing the complete problem and breaking it into its succinct components.  I feel that if we follow something along these lines of working on these problems and coding - especially for the rule-based checker - then it would be seldom that we become surprised later on.

Paul

**Comment by fnothaft** [comment]
@pgrosu that sounds ambitious. Could you put that together and submit it as a follow on PR after https://github.com/ga4gh/schemas/pull/291 merges?

**Comment by macieksmuga** [comment]
@dcolligan the fixes you made to the annotations branch aren't hacks but rather worthy of a PR against the schemas master, if I understand @fnothaft correctly.

**Comment by diekhans** [comment]
Maciek Smuga-Otto <notifications@github.com> writes:
> @dcolligan the fixes you made to the annotations branch aren't hacks but rather
> worthy of a PR against the schemas master, if I understand @fnothaft correctly.

Agreed. a branch off of this and a pull request is in order.

**Comment by fnothaft** [comment]
> @dcolligan the fixes you made to the annotations branch aren't hacks but rather worthy of a PR against the schemas master, if I understand @fnothaft correctly.

+1

**Comment by pgrosu** [comment]
@macieksmuga, @dcolligan Regrading the `ReadAlignment.duplicateFragment` setting, that identification and marking of duplicates is decided by a program after alignment was performed, and it is difficult to decide this sometimes if it should be `true` or `false` for an unmapped read, and sometimes for a RNA-Seq read.  It thus provides flexibility when the call cannot be made confidently.

@fnothaft So in industry usually there is something like a validation environment in terms of setting up the rule-checker/tests/etc, and to set something like that here we would need to expand on the current setup.  This is not an overnight thing, but will provide the framework of checks-and-balances you sometimes see in @dglazer's posts regarding definitions and interpretations.  

As a small example, here is a test-driven process one would go through for just the `Dataset` definition for just the ID generation, given a set of BAM file(s).  This is just an illustrative, simplistic example since the real tests would be more extensive, with a central database-driven server that would provide the correct validation criteria and outputs.  So we know that the `Dataset` definition is the following, and we will just focus on the `id`:

```
record Dataset {
  /** The dataset ID. */
  string id;

  /** The dataset description. */
  union { null, string } description = null;
}
```

*NB: The following process assumes automatic generation of dataset ids from raw data, so that the corresponding Reads and/or Variants would be associated to it.  Other approaches would also be welcomed with the goal of finding the optimal approach.*

First the user who created the `Dataset` PR would provide one or more compressed test-files, which would have the following structure - or something very logically similar:

```
tests/
  Reads /
    Dataset/
      create.dataset-id.USERNAME.3-26-2015
	    Description.txt
	    data/
	      dataset-1/
		    example1.bam
		  dataset-2/
   		    example-1.bam
		    example-2.bam
	    tests/
	      run-test-1.sh
		  run-test-2.sh
```

The `Description.txt` file would contain information regarding the description, scope and interpretation of the test(s).  Within the `data` directory, the `example-1.bam` file would contain something along the following lines - and in this example the `example-2.bam` would be the same file:


```
@HD	VN:1.0	SO:coordinate
@SQ	SN:chr1	LN:101
simple_fits	0	chr1	2	47	100M	*	0	0	TTCATGCTGANGCNCTCTTACGATCGTACAGATGCAAATATTAACANNCNTTNAAGNNCANNNNNNNNNCAATACAATANTAGAGTACGTNAACACTCCA	&/,&-.1/6/&&)&).)/,&0768)&/.,/874,&.4137572)&/&&,&1-&.0/&&*,&&&&&&&&&&18775799,&16:8775-56256/69::;0
```

The `run-test-1.sh` script would be something along these lines, with `run-test-2.sh` pointing to the `data/dataset-2` directory:

```Bash
#!/bin/sh
echo ""
echo "For the following files:"
echo ""
for f in `ls ../data/dataset-1/*.[Bb][Aa][Mm]`; do
  echo "$f"
done
echo ""
echo "The expected md5sum to store as a Dataset id is:"
echo ""
cat ../data/dataset-1/*.[Bb][Aa][Mm] | md5sum
echo
```

Below would be the test results against which to compare, build, and validate any implemented code:

```
$ ./run-test-1.sh

For the following files:

../data/dataset-1/example-1.bam

The expected md5sum to store as a Dataset id is:

303b5e4d072a90cb370268341c403dbd  -

$ ./run-test-2.sh

For the following files:

../data/dataset-2/example-1.bam
../data/dataset-2/example-2.bam

The expected md5sum to store as a Dataset id is:

8830b11c3ad3872ad4662828a1ddb1bb  -

$
```

In this test I utilized `md5sum`, but this can be updated to anything we want.  Again any PR that adds new functionality will need to provide at least one test for every new addition/significant-change to the schema.  In industry, we would would have hundreds - sometimes thousands - of tests that the APIs and implemented codebases would run validity-checks against - including timing differences.  So this is not a one-person job, and to do this right we would together need to put in the time for it, by drawing upon everyone's expertise.  This would be first done for all parts of `schemas` before any code is written, and much of the code design will come from this.  I feel that building using this test-driven design first, will allow us to work in lockstep and will prevent us from being surprised by significantly unexpected results, or from accidentally veering off the path because of presumed interpretations and/or assumptions.  The collection of tests will naturally present the connections and flow in the API, including any major areas of concern that would need to be addressed.  I never saw an instance where this approach did not pay off, and in our case it will ensure that every `schemas` comment - for their respective definition - will always have at least one test behind them.

Let me know what you think.

Thanks,
Paul

**Comment by jeromekelleher** [comment]
Sorry, I'm processing a bit of a backlog here, and trying to figure out what this thread is actually about. WRT to the comments I made in `process_schemas.py`, this was just to highlight that we'd need to deal with the shortcut I'd taken at some point. Are there any outstanding issues *directly germane* to us writing software?

**Comment by dcolligan** [comment]
@jeromekelleher  Mark wanted to create a branch to develop the annotations feature in.  To do this, we needed to base it off of a schemas commit that included the annotation schemas.  I figured we would just use the latest commit in the schemas repository and pushed a branch, schemas/annotations.  Trying to incorporate those schemas into the server code (via `process_schemas.py`), I found a number of bugs.  Others in this thread suggested that I file PRs against the schemas repo to fix those bugs, which I have done.  I have done some work in server/annotations and Mark will apparently be pushing some code in that branch soon.  Currently, the server/annotations branch can create a `_protocol_definitions.py` file successfully, but a lot of tests are failing, and I'm not sure it's worth fixing those tests right now, given the predicted churn in the schemas branch because of iterating on the annotations schemas (but we will need to do that before this branch gets merged into server/develop, of course).  That's where we stand.
